### PR TITLE
Better builtin roundtripping

### PR DESCRIPTION
**Pull request by mahmoud**
glom deals with a lot of callables, and the `repr` of builtin callables like `dict` and `sum` roundtrip poorly. this is an attempt at making and adopting a small cross-platform utility for the benefit of all the glom specifier types that deal with callables.

### CODE REVIEW

**Comment by mahmoud** [comment]
Before: `Assign(T.a, 1, missing=<type 'dict'>)`
After: `Assign(T.a, 1, missing=__builtins__['dict'])`

Unfortunately this varies on the CPython REPL (where `__builtins__` is a module) vs the CPython normal execution (where `__builtins__` is a dict). This PR supports this, but it may be confusing to users.

**Comment by kurtbrose** [comment]
wouldn't you want `Assign(T.a, 1, missing=__builtins__['dict'])` to just be `Assign(T.a, 1, missing=dict])`?

**Comment by mahmoud** [comment]
> wouldn't you want `Assign(T.a, 1, missing=__builtins__['dict'])` to just be `Assign(T.a, 1, missing=dict])`?

Ideally! That's what I started out with. Then arose the issue of the common practice of overriding the `dict` builtin (among others, e.g., `id`). Hence, this explicit compromise.

**Comment by kurtbrose** [comment]
sorry, little confused -- do you mean overriding in builtins, or a global or local with the name "id" or "dict"?

I guess I'm trying to understand the ambiguity?

is it this case?
```python
>>> dict = {}
>>> Assign(T.a, 1)
Assign(T.a, 1, dict)
```

**Comment by kurtbrose** [comment]
```python
>>> a = {}
>>> dict = "user is messing themselves up"
>>> type(a)
<type 'dict'>
>>> type(a) is dict
False
```

**Comment by mahmoud** [comment]
Yeah, I was hoping I could come up with a reasonable explicit way that'll work regardless of what's in scope. But the breakage between dict/module-ness of `__builtins__` has rendered that even more fraught than the shadow-able route. I've reverted that part and now it's just a matter of adopting `bbrepr` internally.

In summary:

Before PR: `Assign(T.a, 1, missing=<type 'dict'>)`
Above: `Assign(T.a, 1, missing=__builtins__['dict'])`
Now: `Assign(T.a, 1, missing=dict)`

File: glom/core.py [hunk 1]
```diff
@@ -32,7 +32,7 @@
 
 from boltons.typeutils import make_sentinel
 from boltons.iterutils import is_iterable
-from boltons.funcutils import format_invocation
+#from boltons.funcutils import format_invocation
```
Comment (kurtbrose) [inline glom/core.py]: is this intended to still be commented out?

File: glom/core.py [hunk 2]
```diff
@@ -248,6 +248,57 @@ def __str__(self):
...
 
 
+if getattr(__builtins__, '__dict__', None) is not None:
+    # pypy's __builtins__ is a module, as is CPython's REPL, but at
+    # normal execution time it's a dict?
+    __builtins__ = __builtins__.__dict__
+
+
+_BUILTIN_ID_NAME_MAP = dict([(id(v), k)
+                             for k, v in __builtins__.items()])
```
Comment (kurtbrose) [inline glom/core.py]: maybe a dict comprehension would be marginally more readable here?

File: glom/core.py [hunk 3]
```diff
@@ -248,6 +248,57 @@ def __str__(self):
...
+    roundtrippable.
+    """
+    ret = repr(obj)
+    if not ret.startswith('<'):
+        return ret
+    return _BUILTIN_ID_NAME_MAP.get(id(obj), ret)
+
+
+# TODO: push this back up to boltons with repr kwarg
+def format_invocation(name='', args=(), kwargs=None, **kw):
```
Comment (kurtbrose) [inline glom/core.py]: oh I see why not importing boltons

**Comment by kurtbrose** [comment]
yea looks good to me tiny tiny suggestion of dict comprehension but lets merge this!

File: glom/core.py (hunk 2 above)
Comment (mahmoud) [reply glom/core.py]: ehhh it's gonna end up in boltons anyways ;)

**Comment by mahmoud** [comment]
Done! Now, to get it working with Match... (should clean up patterns somewhat, but containers like tuples and lists will defeat the change)

**Pull request merged**